### DISCUSSION TITLE
Keep history of each tab separately even after close and reopen tabby terminal

### DISCUSSION THREAD

**Discussion by Kingdom-Of-Heaven** [Eugeny/tabby#8044] [General]
I use tabby terminal and want to save history of each tab so every time my history is kept even after close/open. The good thing is when I write something in one tab and something else on another tab is only kept in those tabs means its good. The problem become when I close tabby and open again history of all tabs seems to be merged. How can I avoid that? This is what I have in my `~/.zshrc`:

  ```
  ...
    source $ZSH/oh-my-zsh.sh
    
    unsetopt inc_append_history
    unsetopt share_history
    ...
```

**Comment by Eugeny** [comment]
This is unfortunately not possible to do 100% automatically - at the end of the day all history is save into `.zsh_history` - your options are just delaying this saving until the shell exits so that at least the _runtime_ histories don't overlap.

To have separate histories, you *must* set separate `$HISTFILE` env vars for each tab .

**Reply by Kingdom-Of-Heaven** [reply]
Hi, How and where exactly should i separate HISTFILE?

**Reply by Kingdom-Of-Heaven** [reply]
This way i almost i would want to have:

```
setopt noincappendhistory
setopt nosharehistory
setopt appendhistory
```

Each tab doesnt see history of another tabs nevertheless once i close my terminal history is merged. Can u guide me how to change that behaviour?

**Reply by Eugeny** [reply]
It's impossible to change unless you make a separate history file for each of your tabs/profiles. ZSH needs to store the history _somewhere_.

**Reply by Kingdom-Of-Heaven** [reply]
and how to make this separate history file?
